raft/src/main/java/org/apache/kafka/raft/utils/BeginQuorumEpochRpc.java

```diff
@@ -0,0 +1,132 @@
+            short apiVersion,
+            Errors topLevelError,
+            TopicPartition topicPartition,
+            Errors partitionLevelError,
+            int leaderEpoch,
+            int leaderId,
+            Endpoints endpoints
+    ) {
+        BeginQuorumEpochResponseData response = new BeginQuorumEpochResponseData()
+                .setErrorCode(topLevelError.code())
+                .setTopics(
+                        Collections.singletonList(
+                                new BeginQuorumEpochResponseData.TopicData()
+                                        .setTopicName(topicPartition.topic())
+                                        .setPartitions(
+                                                Collections.singletonList(
+                                                        new BeginQuorumEpochResponseData.PartitionData()
+                                                                .setErrorCode(partitionLevelError.code())
+                                                                .setLeaderId(leaderId)
+                                                                .setLeaderEpoch(leaderEpoch)
+                                                )
+                                        )
+                        )
+                );
+
+        if (apiVersion >= 1) {
+            Optional<InetSocketAddress> address = endpoints.address(listenerName);
+            if (address.isPresent() && leaderId >= 0) {
+                // Populate the node endpoints
+                BeginQuorumEpochResponseData.NodeEndpointCollection nodeEndpoints =
+                        new BeginQuorumEpochResponseData.NodeEndpointCollection(1);
+                nodeEndpoints.add(
+                        new BeginQuorumEpochResponseData.NodeEndpoint()
+                                .setNodeId(leaderId)
+                                .setHost(address.get().getHostString())
+                                .setPort(address.get().getPort())
+                );
+                response.setNodeEndpoints(nodeEndpoints);
+            }
+        }
+
+        return response;
+    }
+
+    public static boolean hasValidTopicPartition(BeginQuorumEpochRequestData data, TopicPartition topicPartition) {
+        return data.topics().size() == 1 &&
+                data.topics().get(0).topicName().equals(topicPartition.topic()) &&
+                data.topics().get(0).partitions().size() == 1 &&
+                data.topics().get(0).partitions().get(0).partitionIndex() == topicPartition.partition();
+    }
+
+    public static boolean hasValidTopicPartition(BeginQuorumEpochResponseData data, TopicPartition topicPartition) {
+        return data.topics().size() == 1 &&
+                   data.topics().get(0).topicName().equals(topicPartition.topic()) &&
+                   data.topics().get(0).partitions().size() == 1 &&
+                   data.topics().get(0).partitions().get(0).partitionIndex() == topicPartition.partition();
```

Comment:
spacing looks off

Reply:
It seems like an issue of idea format config.
I have adjusted it.